### PR TITLE
exception fixes

### PR DESCRIPTION
**Pull request by akshaydeo**
## Summary

Fixes context propagation for async job execution by ensuring that original request context values (virtual keys, tracing headers, etc.) are properly restored in background goroutines. This resolves issues where async operations would lose important request metadata needed for proper authentication, tracing, and plugin execution.

## Changes

- Modified `AsyncJobExecutor.SubmitJob()` to accept a `contextValues` parameter containing the original request's BifrostContext user values
- Updated `executeJob()` to restore these context values on the background execution context before running the operation
- Added comprehensive test coverage for context propagation scenarios including success, failure, and edge cases
- Fixed streaming response handling to avoid accessing recycled fasthttp RequestCtx objects in goroutines by:
  - Creating snapshot functions that capture request/response data while the context is still valid
  - Implementing `runTransportPostHooksCaptured()` for goroutine-safe plugin execution
  - Updating trace completion callbacks to accept logs as parameters instead of reading from context
- Added `BuildHTTPResponseFromFastHTTP()` utility function for creating response snapshots
- Updated all async handler endpoints to pass context values when submitting jobs

## Type of change

- [x] Bug fix
- [ ] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [x] Transports (HTTP)
- [ ] Providers/Integrations
- [ ] Plugins
- [ ] UI (Next.js)
- [ ] Docs

## How to test

Validate that async operations properly inherit request context:

```sh
# Run the new async job tests
go test ./framework/logstore -v -run TestSubmitJob

# Test full async workflow with context propagation
go test ./transports/bifrost-http/handlers -v -run TestAsync

# Verify no race conditions in streaming responses
go test -race ./transports/bifrost-http/handlers

# Run full test suite
go test ./...
```

Test async endpoints with virtual keys and custom headers to ensure context values are preserved in background execution.

## Screenshots/Recordings

N/A

## Breaking changes

- [ ] Yes
- [x] No

## Related issues

Addresses context loss in async job execution that could cause authentication failures and incomplete tracing data.

## Security considerations

Improves security by ensuring virtual key validation and authentication context is properly maintained in async operations. No new security risks introduced as this only preserves existing context data.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [ ] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by akshaydeo** [comment]
* **#2676** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2676?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2676?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `v1.5.0`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: transports/bifrost-http/handlers/inference.go [hunk 1]
```diff
@@ -1667,8 +1668,17 @@ func (h *CompletionHandler) handleStreamingResponse(ctx *fasthttp.RequestCtx, bi
...
 	ctx.SetUserValue(schemas.BifrostContextKeyDeferTraceCompletion, true)
 
-	// Get the trace completer function for use in the streaming callback
-	traceCompleter, _ := ctx.UserValue(schemas.BifrostContextKeyTraceCompleter).(func())
+	// Pre-allocate atomic.Value slot for the transport post-hook completer.
+	// TransportInterceptorMiddleware stores the completer into this slot after next(ctx)
+	// returns. The goroutine reads from the closure-captured pointer, avoiding any ctx
+	// access after the handler returns (fasthttp recycles RequestCtx).
+	var completerSlot atomic.Value
+	ctx.SetUserValue(schemas.BifrostContextKeyTransportPostHookCompleter, &completerSlot)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/inference.go:1677]: Fixed in latest push. Added a `runCompleter` closure in `inference.go` that polls `completerSlot.Load()` with a 100ms deadline (1ms sleep) before giving up. The wait is well above the middleware's post-`next(ctx)` store latency in normal scheduling; on timeout we fall through to the trace completer with empty transport logs instead of silently dropping the post-hook. Combined with the fix for the adjacent comment so the completer also runs before `[DONE]`. Can you check and resolve if done properly?

File: transports/bifrost-http/handlers/inference.go [hunk 2]
```diff
@@ -1686,23 +1696,28 @@ func (h *CompletionHandler) handleStreamingResponse(ctx *fasthttp.RequestCtx, bi
...
+			// ctx may already be recycled by fasthttp at this point.
+			var transportLogs []schemas.PluginLogEntry
+			if loaded := completerSlot.Load(); loaded != nil {
+				if postHookCompleter, ok := loaded.(func() ([]schemas.PluginLogEntry, error)); ok {
+					logs, err := postHookCompleter()
+					if err != nil {
+						errorJSON, marshalErr := sonic.Marshal(map[string]string{"error": err.Error()})
+						if marshalErr == nil {
+							reader.SendError(errorJSON)
+						}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/inference.go]: Fixed in latest push. The post-hook completer now runs on the happy path **before** `reader.SendDone()`, so plugin errors are delivered as `event: error` SSE frames the client can observe. The defer still runs a fallback invocation for early-return paths (client disconnect, interceptor error), but in that mode errors are logged server-side via `logger.Warn` instead of `reader.SendError` — the stream is already terminated at that point. A `completerRan` guard prevents double execution. Can you check and resolve if done properly?

File: transports/bifrost-http/handlers/middlewares.go [hunk 3]
```diff
@@ -427,6 +460,58 @@ func runTransportPostHooks(ctx *fasthttp.RequestCtx, plugins []schemas.HTTPTrans
...
+	for k, v := range capturedReq.PathParams {
+		req.PathParams[k] = v
+	}
+
+	httpResp := schemas.AcquireHTTPResponse()
+	defer schemas.ReleaseHTTPResponse(httpResp)
+	httpResp.StatusCode = capturedResp.StatusCode
+	for k, v := range capturedResp.Headers {
+		httpResp.Headers[k] = v
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/middlewares.go:489]: Leaving as-is by design. `capturedReq.Body` and `capturedResp.Body` are intentionally empty in this path:

- `BuildHTTPRequestFromFastHTTP` (`transports/bifrost-http/lib/ctx.go:580`) explicitly skips body with an in-code comment: *"Body not copied — for streaming, body was already consumed."*
- `BuildHTTPResponseFromFastHTTP` (`lib/ctx.go:619`) explicitly skips body: *"body is skipped because for streaming responses it is an active io.Reader that cannot be materialized."*

fasthttp does not permit reading the response body once `SetBodyStream` is active — doing so drains the live SSE stream and corrupts the actual response going to the client. The suggested `if len(capturedReq.Body) > 0 { append(...) }` would compile but never execute (the snapshots never carry bytes), and implies a guarantee the transport layer does not provide.

Verified no `HTTPTransportPostHook` implementation currently inspects `req.Body` / `resp.Body`. Governance reads `req.Body` only in `PreLLMHook`, which is a different plugin phase that runs inline, not on the deferred streaming path. Can you resolve this thread?

File: transports/bifrost-http/handlers/middlewares.go [hunk 4]
```diff
@@ -936,10 +1021,11 @@ func (m *TracingMiddleware) Middleware() schemas.BifrostHTTPMiddleware {
...
 
-			// Store a trace completion callback for streaming handlers to use
-			ctx.SetUserValue(schemas.BifrostContextKeyTraceCompleter, func() {
-				// Attach transport plugin logs before completing the trace (streaming path)
-				if transportLogs, ok := ctx.UserValue(schemas.BifrostContextKeyTransportPluginLogs).([]schemas.PluginLogEntry); ok && len(transportLogs) > 0 {
+			// Store a trace completion callback for streaming handlers to use.
+			// Accepts transport plugin logs as a parameter so it never reads from
+			// ctx.UserValue — ctx may be recycled by the time this runs in a goroutine.
+			ctx.SetUserValue(schemas.BifrostContextKeyTraceCompleter, func(transportLogs []schemas.PluginLogEntry) {
+				if len(transportLogs) > 0 {
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/middlewares.go:1028]: Fixed in latest push. Two edits:

1. `transports/bifrost-http/integrations/router.go:2318` — type assertion updated to `func([]schemas.PluginLogEntry)`. The router streaming path has no transport post-hook phase (no `completerSlot` mechanism wired up here — grep confirmed), so the call site at :2341 passes `nil` for transport logs.
2. `core/schemas/bifrost.go:215` — the key's trailing doc comment updated from `// func()` to `// func([]PluginLogEntry)` so the schema docs match the runtime signature.

Verified with `grep 'BifrostContextKeyTraceCompleter.*func\(\)'` — zero matches remain. Build passes. Can you check and resolve if done properly?

**Comment by akshaydeo** [comment]
@coderabbitai review

Pushed fixes for the 4 unresolved threads from the previous review:
- `3072425551` — race on `completerSlot` → added bounded 100ms poll in `inference.go` `runCompleter` closure
- `3072425564` — SSE error after `[DONE]` → completer now runs before `reader.SendDone()` on happy path; defer fallback logs server-side only
- `3072425566` — captured body copy → left as-is by design (snapshots intentionally skip body; fasthttp `SetBodyStream` forbids reading response body; no post-hook plugin reads body)
- `3072425594` — router.go trace completer signature drift → fixed type assertion + updated schema doc comment in `core/schemas/bifrost.go:215`

Please re-review and resolve the threads if satisfied.

File: transports/bifrost-http/handlers/inference.go [hunk 5]
```diff
@@ -1684,25 +1695,64 @@ func (h *CompletionHandler) handleStreamingResponse(ctx *fasthttp.RequestCtx, bi
...
+					break
+				}
+				if time.Now().After(deadline) {
+					break
+				}
+				time.Sleep(time.Millisecond)
+			}
+			if loaded == nil {
+				return
+			}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/inference.go:1725]: Fixed — moved `completerRan = true` to after the nil-check and type assertion so the deferred fallback at inference.go:1750 can still invoke the completer if the 100ms wait expired with an empty slot. Can you verify and resolve?

File: transports/bifrost-http/handlers/middlewares.go [hunk 6]
```diff
@@ -357,10 +357,43 @@ func TransportInterceptorMiddleware(config *lib.Config) schemas.BifrostHTTPMiddl
...
+
+				completer := func() ([]schemas.PluginLogEntry, error) {
+					defer schemas.ReleaseHTTPRequest(capturedReq)
+					defer schemas.ReleaseHTTPResponse(capturedResp)
+					postHookLogs, err := runTransportPostHooksCaptured(capturedReq, capturedResp, plugins, bifrostCtx)
+					allLogs := preHookLogs
+					if len(postHookLogs) > 0 {
+						allLogs = append(allLogs, postHookLogs...)
+					}
+					return allLogs, err
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/middlewares.go:390]: Partially fixed, partially replying:

**Error string case:** Fixed — lowercased the prefix at middlewares.go:505 and also brought the identical pre-existing string at :446 into compliance (`HTTPTransportPostHook plugin %s: %w` → `transport post-hook plugin %s: %w`).

**Context delegation to recycled parent:** Not making changes here. In the current flow, the deferred completer is always invoked from the streaming goroutine's defer **before** `reader.Done()` (inference.go:1750 → 1751). Since `reader.Done()` is just `close(eventCh)` (streamreader.go:112-114), fasthttp only sees EOF on its next Read and recycles the RequestCtx **after** that. So when `bifrostCtx.parent.Value(k)` delegates to the parent, it reads a live `fasthttp.RequestCtx`. The eager snapshotting of `capturedReq`/`capturedResp`/`preHookLogs` is defense-in-depth in case the ordering is ever changed — which is also what the `IMPORTANT:` comment at middlewares.go:360-363 warns future editors against. Happy to revisit if you can identify a path where the completer fires after `reader.Done()`.

Can you verify and resolve?

File: transports/bifrost-http/handlers/middlewares.go [hunk 7]
```diff
@@ -427,6 +460,58 @@ func runTransportPostHooks(ctx *fasthttp.RequestCtx, plugins []schemas.HTTPTrans
...
+		pluginName := plugin.GetName()
+		pluginCtx := bifrostCtx.WithPluginScope(&pluginName)
+		err := plugin.HTTPTransportPostHook(pluginCtx, req, httpResp)
+		pluginCtx.ReleasePluginScope()
+		if err != nil {
+			logger.Warn("error in HTTPTransportPostHook for plugin %s: %s", pluginName, err.Error())
+			if postHookLogs := bifrostCtx.DrainPluginLogs(); len(postHookLogs) > 0 {
+				allLogs = append(allLogs, postHookLogs...)
+			}
+			return allLogs, fmt.Errorf("HTTPTransportPostHook plugin %s: %w", pluginName, err)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/middlewares.go]: Fixed — lowercased the prefix at middlewares.go:505 (`HTTPTransportPostHook plugin %s: %w` → `transport post-hook plugin %s: %w`). Also applied the same fix to the identical pre-existing string at :446 for consistency. Can you verify and resolve?
